<?xml version="1.0"?>
<launch>

  <arg name="verbose"              default="false"                                   doc="Should more debug output be printed?" />
  <arg name="unpause"              default="false"                                   doc="Should the simulation be unpaused on start?" />
  <arg name="modelfile"            default="$(find mujoco_ros)/test/empty_world.xml" doc="MuJoCo xml file to load. Should define robot model and world." />
  <arg name="headless"             default="false"                                   doc="Should the simulation be rendered." />
  <arg name="num_sim_steps"        default="-1"                                      doc="Number of simulation steps to run. Set to '-1' to run indefinitely." />
  <arg name="use_sim_time"                                                           doc="Should mujoco ublish its simulation time as ros time?" />
  <arg name="wait_for_xml"         default="false"                                   doc="Should mujoco_ros wait for a modelfile on the parameter server" />
  <arg name="mujoco_plugin_config" default=""                                        doc="Optionally provide the path to a yaml with plugin configurations to load" />
  <arg name="realtime"             default="1.0"                                     doc="Desired realtime factor. Set to '-1' to run as fast as possbile " />
  <arg name="render_offscreen"     default="true"                                    doc="Should camera streams be rendered and published?" />
  <arg name="debug"                default="false"                                   doc="Should mujoco be started in gdb?" />
  <arg name="eval_mode"            default="false"                                   doc="Should mujoco be started in evaluation mode?" />
  <arg name="admin_hash"           default="''"                                      doc="Hash to verify critical operations in evaluation mode" />

  <param name="/use_sim_time" value="$(arg use_sim_time)"/>

  <env name="ROSCONSOLE_FORMAT" value="[${severity}] [${time}] [${logger}] [${node}]: ${message}"/>
  <env if="$(arg verbose)" name="ROSCONSOLE_CONFIG_FILE"
       value="$(find mujoco_ros)/rosconsole.config"/>

  <!-- pluign loading -->
  <!-- If MujocoRosControlPlugin is built, but no urdf is uploaded the simulation will block until it is, thus this is commented -->
  <!-- <rosparam>
    MujocoPlugins:
      - type: mujoco_ros_control/MujocoRosControlPlugin
        hardware:
          type: mujoco_ros_control/DefaultRobotHWSim
          control_period: 0.001
  </rosparam> -->
  <group if="$(eval arg('mujoco_plugin_config') != '')">
    <rosparam file="$(arg mujoco_plugin_config)" subst_value="true" />
  </group>

  <group if="$(arg debug)">
    <node pkg="mujoco_ros" type="mujoco_server" name="mujoco_server" output="screen" launch-prefix="gdb -ex run --args" args="--admin-hash $(arg admin_hash)" >
      <param name="modelfile" type="str"  value="$(arg modelfile)" />
      <param name="unpause"   value="$(arg unpause)" />
      <param name="visualize" value="$(eval not arg('headless'))" />
      <param name="num_steps" value="$(arg num_sim_steps)" />
      <param name="realtime"  value="$(arg realtime)" />
      <param name="eval_mode"  value="$(arg eval_mode)" />
    </node>
  </group>
  <group unless="$(arg debug)">
    <node pkg="mujoco_ros" type="mujoco_server" name="mujoco_server" output="screen" args="--admin-hash $(arg admin_hash)" >
      <param name="modelfile" type="str"  value="$(arg modelfile)" />
      <param name="unpause"   value="$(arg unpause)" />
      <param name="visualize" value="$(eval not arg('headless'))" />
      <param name="num_steps" value="$(arg num_sim_steps)" />
      <param name="realtime"  value="$(arg realtime)" />
      <param name="eval_mode"  value="$(arg eval_mode)" />
    </node>
  </group>
</launch>
